### PR TITLE
Add a 'remacs-modules' crate for building Emacs/Remacs modules in Rust.

### PR DESCRIPTION
**Pull request by DavidDeSimone**
I recently verified https://github.com/Wilfred/remacs/issues/464, by writing a small basic C module. It worked, but it got me thinking about writing emacs/remacs modules in Rust. In order to make this easier, I made a Rust crate that wraps the types provided by emacs-modules.h. The idea is that this crate can be used by emacs modules as a dependency. The good news is that these modules should be usable by both emacs AND remacs, so this can be useful for people who want to write modules in Rust but want them to work for both platforms. 

I decided to use `bindgen` to generate the code for emacs-modules.h. I like the idea of using bindgen where we can, and emacs-modules.h is fairly self contained. 

Some notes:

On OSX, Cargo produces .dylibs. By default, the shared_lib extension in remacs is .so on OSX. This can be changed, or you can work around it by manually calling (module-load "\<name\>"), but it is a little annoying.

### CODE REVIEW

File: rust_src/remacs-modules/lib.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        let mut args = vec![qfeat];
+        self.funcall(qprovide, &mut args);
+    }
+
+    pub fn funcall(&mut self, fun: EmacsValue, args: &mut Vec<EmacsValue>) -> EmacsValue {
+        let fnptr = self.funcall.unwrap();
+        let len = args.len() as libc::ptrdiff_t;
+        let mut raw_array: Vec<emacs_value> = args.iter().map(EmacsValue::to_raw).collect();
+        EmacsValue(unsafe { (fnptr)(self.0, fun.to_raw(), len, raw_array.as_mut_ptr()) })
+
```
Comment (shaleh) [inline rust_src/remacs-modules/lib.rs]: empty line

File: rust_src/remacs-sys/build.rs [hunk 2]
```diff
@@ -196,7 +197,36 @@ fn generate_globals() {
...
+
+    let bindings = builder
+        .rustfmt_bindings(true)
+        .generate()
+        .expect("Unable to generate bindings");
+
+    let source = bindings.to_string();
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap()).join("modules.rs");
+    let file = File::create(out_path);
+    file.unwrap().write_all(source.as_bytes()).unwrap();
```
Comment (shaleh) [inline rust_src/remacs-sys/build.rs]: Checkout the error handling I added to the top level build.rs. If this function returned a `Result` with `io::Error` you could use `?` handling to clarify this logic and avoid `unwrap` failures.

File: rust_src/remacs-modules/lib.rs [hunk 3]
```diff
@@ -1,11 +1,14 @@
-extern crate remacs_sys;
 extern crate libc;
```
Comment (shaleh) [inline rust_src/remacs-modules/lib.rs]: Spacing and grouping matters. std, then external, then remacs, then local is the pattern the rest of the code follows.

**Comment by DavidDeSimone** [comment]
Removing WIP tag from this.

File: rust_src/remacs-modules/lib.rs [hunk 4]
```diff
@@ -0,0 +1,363 @@
+extern crate libc;
+extern crate remacs_sys;
+
+use remacs_sys::{emacs_env, emacs_runtime, emacs_value};
```
Comment (shaleh) [inline rust_src/remacs-modules/lib.rs]: Grouping here.

File: rust_src/remacs-modules/lib.rs [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+pub type intmax_t = ::libc::intmax_t;
+
+struct UserData {
+    func: Box<FnMut(&mut EmacsEnv, &mut Vec<EmacsValue>) -> EmacsValue>,
+}
+
+macro_rules! call {
+    ($self: ident, $func: ident) => (call!($self, $func,));
+    ($self: ident, $func: ident, $($arg:expr),*) => {{
+        let fnptr = $self.$func.unwrap();
```
Comment (shaleh) [inline rust_src/remacs-modules/lib.rs]: Do we want to handle errors on unwrap here?

**Comment by ubolonton** [comment]
This is somewhat similar to https://github.com/ubolonton/emacs-module-rs. Perhaps we can combine the efforts somehow?

There are 2 big differences I noticed:
- This uses a single C function that routes to the corresponding Rust function based on data in the extra pointer. My version uses macro to generate different C functions. I'm not sure which approach is more favorable.
- My version converts between non-local exit and `Result` at the Emacs-module boundary. Module functions deal only with `Result`. Any thought on whether this is an appropriate strategy?

The most unclear issue to me is the ownership and lifetime of `EmacsValue`. From this doc https://phst.github.io/emacs-modules (the best one I found so far on the topic), it seems like:
- The lifetime of `EmacsValue` should be the that of the associated `EmacsEnv`. Module functions should only get `&mut EmacsValue`.
- `EmacsValue` should not implement `Clone`.
   + A special function `env.clone(&value)` should be used to get an owned version instead (which should call `make_global_ref`)
   + Their `Drop` implementation should call `free_global_ref`.

**Comment by shaleh** [comment]
@ubolonton sorry for not seeing this earlier. I was deferring to @DavidDeSimone.

We have currently decided not to use `Result` in the remacs port because it disrupts the lisp stack when a failure occurs. With `result` the failure is no longer reported at the point it occurred but at the point it was noticed. Which is different from using `signal!`. If you look at the remacs code where you have `Result` as the return value from a `From` trait we use `Option` if the cast can fail or we explicitly use `as_foo_or_error()` which calls `signal!`. Which is similar to invoking `unwrap` on a `Result` that is not `Ok`.

I need to examine the code more closely and ponder before I can comment on whether the macro or all-in-one approach is better.

Your reading about the lifetimes matches my expectation.

**Comment by shaleh** [comment]
I am closing this as it is getting more and more out of date.

@ubolonton if you have interest please open a new PR.

**Pull request closed without merging**